From 6e3c4449ae333411349769b566bf1f1c7a803e8c Mon Sep 17 00:00:00 2001
From: Subrata Chatterjee <schatterjee@habana.ai>
Date: Tue, 08 Sep 2020 13:41:15 +0300
Subject: [PATCH] [SW-17451] goya2: led: add led control

Yellow status LED(GPIO2) is turned on when preboot
is up.

Now status LED control sequence would be as follows,
 -Preboot up: Yellow LED on
 -Zephyr on ARCs up: Green LED on, Yellow led already on
 -Linux up: Green LED on(already done by zephyr sw),
	    Yellow LED off

Change-Id: I9b58564aaf28d084a88eefe3d507809da2f9778e
Signed-off-by: Subrata Chatterjee <schatterjee@habana.ai>
---

diff --git a/goya2/asic.c b/goya2/asic.c
index 37f3477..cf0f54a 100644
--- a/goya2/asic.c
+++ b/goya2/asic.c
@@ -6,6 +6,7 @@
  */
 
 #include <psoc_global_conf_regs.h>
+#include <psoc_gpio0_regs.h>
 
 #include <preboot.h>
 #include <pll.h>
@@ -48,3 +49,17 @@
 
 	return factor > 2 ? factor : 2;
 }
+
+/* configure yellow LED(GPIO2) in output mode and turn on(active low) */
+void preboot_led_on(void)
+{
+	u32 val = BIT(2);
+
+	/* configure GPIO in output mode */
+	regmap_writel(val, mmPSOC_GPIO0_OUTENSET);
+
+	/* Turn ON LED (pin is active low) */
+	val = regmap_readl(mmPSOC_GPIO0_DATAOUT);
+	val &= ~BIT(2);
+	regmap_writel(val, mmPSOC_GPIO0_DATAOUT);
+}
diff --git a/goya2_arc_gnu/asic.c b/goya2_arc_gnu/asic.c
index f54e71b..6d3fa5e 100644
--- a/goya2_arc_gnu/asic.c
+++ b/goya2_arc_gnu/asic.c
@@ -6,6 +6,7 @@
  */
 
 #include <psoc_global_conf_regs.h>
+#include <psoc_gpio0_regs.h>
 
 #include <preboot.h>
 #include <pll.h>
@@ -31,3 +32,17 @@
 
 	preboot_uart_init(uart_clk_freq / HL_UART_BAUDRATE);
 }
+
+/* configure yellow LED(GPIO2) in output mode and turn on(active low) */
+void preboot_led_on(void)
+{
+	u32 val = BIT(2);
+
+	/* configure GPIO in output mode */
+	regmap_writel(val, mmPSOC_GPIO0_OUTENSET);
+
+	/* Turn ON LED (pin is active low) */
+	val = regmap_readl(mmPSOC_GPIO0_DATAOUT);
+	val &= ~BIT(2);
+	regmap_writel(val, mmPSOC_GPIO0_DATAOUT);
+}
diff --git a/include/preboot.h b/include/preboot.h
index e2ab612..010a7ae 100644
--- a/include/preboot.h
+++ b/include/preboot.h
@@ -109,6 +109,7 @@
 void preboot_putc_virt_uart(const char c);
 void preboot_sync_btl(void);
 void preboot_security_init(void);
+void preboot_led_on(void);
 
 void ppboot_cp_preboot(u64 loadaddr, u32 offset, u32 size, char *name);
 void image_start(u64 loadaddr);
diff --git a/preboot/preboot.c b/preboot/preboot.c
index 0e3cef4..6c2d60c 100644
--- a/preboot/preboot.c
+++ b/preboot/preboot.c
@@ -110,6 +110,7 @@
 }
 
 __weak void preboot_security_init(void) {}
+__weak void preboot_led_on(void) {}
 
 void preboot_main(void)
 {
@@ -130,5 +131,6 @@
 
 	hl_ts_init();
 	preboot_version();
+	preboot_led_on();
 	module_handle_cmds(cmd_ops, ARRAY_SIZE(cmd_ops));
 }
